packages/flutter/lib/src/rendering/wrap.dart

```diff
@@ -747,25 +794,56 @@ class RenderWrap extends RenderBox
 
     _RunMetrics? currentRun;
     _AxisSize childrenAxisSize = _AxisSize.empty;
+
     for (RenderBox? child = firstChild; child != null; child = childAfter(child)) {
       final _AxisSize childSize = _AxisSize.fromSize(
         size: layoutChild(child, childConstraints),
         direction: direction,
       );
-      final _RunMetrics? newRun = currentRun == null
-          ? _RunMetrics(child, childSize)
-          : currentRun.tryAddingNewChild(child, childSize, flipMainAxis, spacing, mainAxisLimit);
-      if (newRun != null) {
-        runMetrics.add(newRun);
-        childrenAxisSize += currentRun?.axisSize.flipped ?? _AxisSize.empty;
-        currentRun = newRun;
+
+      final bool needsNewRun =
+          currentRun == null ||
+          currentRun.axisSize.mainAxisExtent + childSize.mainAxisExtent + spacing > mainAxisLimit;
+
+      // If we need a new run and are about to exceed maxLines, stop.
+      if (needsNewRun && currentRun != null) {
```

Comment:
I have some thoughts about simplifying this logic a bit, I think it can be streamlined to not have as many breaks and continue, but I think we need to account for the edge cases like 0 first. :)